src/main/java/org/gaul/s3proxy/junit/S3ProxyRule.java

```diff
@@ -0,0 +1,189 @@
+import org.gaul.s3proxy.AuthenticationType;
+import org.gaul.s3proxy.S3Proxy;
+import org.jclouds.ContextBuilder;
+import org.jclouds.blobstore.BlobStore;
+import org.jclouds.blobstore.BlobStoreContext;
+import org.jclouds.blobstore.domain.StorageMetadata;
+import org.junit.rules.ExternalResource;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.annotations.Beta;
+
+/**
+ * A JUnit Rule that manages an S3Proxy instance which tests can use as an S3 API endpoint.
+ */
+@Beta
+public final class S3ProxyRule extends ExternalResource {
+    private static final Logger logger = LoggerFactory.getLogger(
+        S3ProxyRule.class);
+
+    private static final String LOCALHOST = "127.0.0.1";
+
+    private final String accessKey;
+    private final String secretKey;
+    private final String endpointFormat;
+    private final S3Proxy s3Proxy;
+
+    private BlobStoreContext blobStoreContext;
+    private URI endpointUri;
+
+    public static final class Builder {
+        private AuthenticationType authType = AuthenticationType.NONE;
+        private String accessKey;
+        private String secretKey;
+        private String secretStorePath;
+        private String secretStorePassword;
+        private int port = -1;
+        private boolean ignoreUnknownHeaders;
+        private String blobStoreProvider = "filesystem";
+
+        private Builder() { }
+
+        public Builder withCredentials(AuthenticationType authType,
+                String accessKey, String secretKey) {
+            this.authType = authType;
+            this.accessKey = accessKey;
+            this.secretKey = secretKey;
+            return this;
+        }
+
+        public Builder withCredentials(String accessKey, String secretKey) {
+            return withCredentials(AuthenticationType.AWS_V2_OR_V4, accessKey,
+                secretKey);
+        }
+
+        public Builder withSecretStore(String path, String password) {
+            secretStorePath = path;
+            secretStorePassword = password;
+            return this;
+        }
+
+        public Builder withPort(int port) {
+            this.port = port;
+            return this;
+        }
+
+        public Builder withBlobStoreProvider(String blobStoreProvider) {
+            this.blobStoreProvider = blobStoreProvider;
+            return this;
+        }
+
+        public Builder ignoreUnknownHeaders() {
+            ignoreUnknownHeaders = true;
+            return this;
+        }
+
+        public S3ProxyRule build() {
+            return new S3ProxyRule(this);
+        }
+    }
+
+    private S3ProxyRule(Builder builder) {
+        accessKey = builder.accessKey;
+        secretKey = builder.secretKey;
+
+        Properties properties = new Properties();
+        try {
+            File blobStoreLocation = Files.createTempDirectory("S3ProxyRule")
+                    .toFile();
+            blobStoreLocation.deleteOnExit();
```

Comment:
Remove `deleteOnExit` now that you explicitly remove files in `after`?

Reply:
I tried this earlier, if you don't do this then the top-level temporary folder is left behind after the test runs.

Reply:
I see, if you need to remove the top-level directory, do so explicitly in `after`.  Relying on `deleteOnExit` is fragile.

Reply:
This is why we were using a JUnit temporary folder in before() and after(). Can I please put that code back? Then we don't need the blob store context and its cleanup any more either.

Reply:
Explicitly clearing the `BlobStore` allows cleaning up non-filesystem blobstores.  We should actually guard the creation and deletion of the temporary folder in an `if (blobStoreProvider.equals("filesystem))` check.  Using the JUnit mechanism unnecessarily creates and removes an unnecessary directory for the transient blobstore.

Reply:
But we're only ever using a filesystem in the rule...

Reply:
But the code allows changing the provider via `withBlobStoreProvider`.  Using the transient blobstore has slightly better S3 compatibility than filesystem historically.

Reply:
OK, added a `deleteQuietly` in `after()` instead of `deleteOnExit()`. Confirmed that this also works.